### PR TITLE
Add types for "next-auth' package 📥

### PR DESCRIPTION
**Pull request by 0ubbe**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disable of rules
- [x] `tsconfig.json` has `strict` set to `true`.

### CODE REVIEW

File: types/next-auth/tsconfig.json
Comment (arnodel) [inline types/next-auth/tsconfig.json:17]: I believe the build fails to parse this file because of the trailing comma on this line

File: types/next-auth/client.d.ts
Comment (NickBolles) [inline types/next-auth/client.d.ts:54]: I think this is incorrect. The actual export is an object exported as default, from what I understand, this is exporting each of the functions directly.

https://github.com/nextauthjs/next-auth/blob/main/src/client/index.js#L323-L343

**Comment by NickBolles** [comment]
hey @lluia I just put together https://github.com/nextauthjs/next-auth/pull/516. While doing that I found the following issues with these typings. Could you make the updates here?

1. Missing Providers typings (these would not show up for typescript users at all)
  - Linked In 
  - Spotify
  - Reddit
2. Email config Typings
  - It looks like these two were on the ProviderEmailServer, when it looks like they should be on ProviderEmailOptions (although I'm not too sure what ProviderEmailServer should be)
```ts
maxAge: number;
sendVerificationRequest: typeof sendVerificationRequest;
```
3. jwt on InitOptions shouldn't be a boolean
  - https://next-auth.js.org/configuration/options#jwt
  - and it will crash if it's a boolean at runtime
 ```ts
// JWT options
    const jwtOptions = {
      secret, // Use application secret if no keys specified
      maxAge: sessionOptions.maxAge, // maxAge is dereived from session maxAge,
      encode: jwt.encode,
      decode: jwt.decode,
      ...userSuppliedOptions.jwt
    }
```
4. `PageOptions.signin` should be `PageOptions.signIn` (capital "i")
5. `PageOptions.newUsers` should be `PageOptions.newUser` (no "s")
6. Fix "for" JSX attribute
 ```bash
src/server/pages/signin.tsx:67:22 - error TS2322: Type '{ children: string; for: string; }' is not assignable to type 'DetailedHTMLProps<LabelHTMLAttributes<HTMLLabelElement>, HTMLLabelElement>'.
        Property 'for' does not exist on type 'DetailedHTMLProps<LabelHTMLAttributes<HTMLLabelElement>, HTMLLabelElement>'.

      67               <label for={`input-email-for-${provider.id}-provider`}>Email</label>
                              ~~~
```
and
```
      src/server/pages/signin.tsx:78:23 - error TS2322: Type '{ children: any; for: string; }' is not assignable to type 'DetailedHTMLProps<LabelHTMLAttributes<HTMLLabelElement>, HTMLLabelElement>'.
        Property 'for' does not exist on type 'DetailedHTMLProps<LabelHTMLAttributes<HTMLLabelElement>, HTMLLabelElement>'.

      78                       for={`input-${credential}-for-${provider.id}-provider`}
```

File: types/next-auth/client.d.ts
Comment (lluia) [reply types/next-auth/client.d.ts:54]: Good catch 🧐 💯 , as you mention it's a _default export_ with the functions on it, I'll amend 👍🏻

**Comment by lluia** [comment]
@NickBolles awesome work 💯 💚 , I'm gonna update all the missing things you found 👌🏻

File: types/next-auth/client.d.ts
Comment (rbuckton) [reply types/next-auth/client.d.ts]: > I think this is incorrect. The actual export is an object exported as default, from what I understand, this is exporting each of the functions directly.

Its actually not, unfortunately. According to unpkg, [client.js](https://unpkg.com/browse/next-auth@3.1.0/client.js) looks like this:

```js
module.exports = require('./dist/client').default
```

So you end up with the module itself actually being `{ useSession, getSession, ... }` instead of a default export.

Comment (NickBolles) [reply types/next-auth/client.d.ts]: @rbuckton, you're right. I was looking at `client/index.js` instead of `client.js` @lluia, my bad, could you switch it back?

File: types/next-auth/client.d.ts
Comment (rbuckton) [inline types/next-auth/client.d.ts:76]: Per https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45775/files/ef8af8c4ce9c2ce40dcced2e35dcb61ff909e941#r466094897, these should be exported via `export { useSession, ... }` as was provided previously, rather than via `export default`.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/next-auth/client.d.ts
Comment (lluia) [reply types/next-auth/client.d.ts]: Yup I'll update 👍

File: types/next-auth/client.d.ts
Comment (lluia) [reply types/next-auth/client.d.ts]: @rbuckton addressed 👍🏽

**Comment by andrewbranch** [comment]
@NickBolles would you mind double checking if all your concerns have been addressed? I’ll merge as soon as I see an approval from someone familiar with next-auth. Thanks!

File: types/next-auth/client.d.ts
Comment (adambrgmn) [inline types/next-auth/client.d.ts:9]: Access token is not always available on the session object – only if `session.jwt` is set to `false` on the configuration. So this should probably be:

```suggestion
    accessToken?: string;
```

Comment (adambrgmn) [reply types/next-auth/client.d.ts:9]: https://github.com/nextauthjs/next-auth/blob/2ea64045cb28cbe0167a9bd8cd65c2d752e452fb/src/server/routes/session.js#L19

**Review by adambrgmn — COMMENTED** [review]
Nice I was thinking about making this job myself, but you've done a great job! Thank you very much for that 👍 

Just a few questions below.

File: types/next-auth/index.d.ts
Comment (lluia) [reply types/next-auth/index.d.ts]: Actually gonna revert this is it [throws on the CI](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45775/checks?check_run_id=990581452) related to [this rule](https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md) 💥

**Comment by DanielRosenwasser** [comment]
I'll merge this tomorrow unless there's objections

File: types/next-auth/index.d.ts
Comment (SMores) [inline types/next-auth/index.d.ts:14]: Would it be possible to export this interface? I'm trying these types out locally on a new project using NextAuth, and for reasons that I don't totally grok, the TS compiler seems to struggle to infer the correct `ConnectionOptions` type if I don't define my `options` object as being of type `InitOptions` up front. That is, this doesn't work:

```typescript
const options = {
  database: {
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT, 10),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
  },
}

export default (req, res) => NextAuth(req, res, options)
```

Because the compiler complains that:

> Type '{ type: string; host: string; port: number; username: string; password: string; database: string; }' is missing the following properties from type 'AuroraDataApiConnectionOptions': region, secretArn, resourceArn

But this is fine: 

```typescript
const options: InitOptions = {
  database: {
    type: 'postgres',
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT, 10),
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
  },
}

export default (req, res) => NextAuth(req, res, options)
```

Comment (lluia) [reply types/next-auth/index.d.ts:14]: @SMores yup definitively fine having those ones exported 👍🏽 , let me amend this in the afternoon 🗓

Comment (lluia) [reply types/next-auth/index.d.ts]: @SMores addressed, if you can  👀 again plz

**Comment by daviduzumeri** [comment]
Is there a reason that `signIn` and `signOut` have required parameters? (Also trying these out locally in a new project, so thank you so much!) Just wondering since the examples on the `next-auth` website all specify that they're optional (see: https://next-auth.js.org/getting-started/client#signin) and also that a string can be used as a `SessionProvider`, but these typings seem to require objects as session providers and for both parameters to be presented.

**Comment by lluia** [comment]
@daviduzumeri good catch 👌🏽 , I'll try to update those soon 👍🏽

**Comment by daviduzumeri** [comment]
> @daviduzumeri good catch 👌🏽 , I'll try to update those soon 👍🏽

Hey, thank you for taking this on! Next-auth is a wildly dope package and I'm excited to get full TS compat with it!

**Comment by daviduzumeri** [comment]
> > @daviduzumeri good catch 👌🏽 , I'll try to update those soon 👍🏽
> 
> Hey, thank you for taking this on! Next-auth is a wildly dope package and I'm excited to get full TS compat with it!

(Another thing: It looks like the `req` on `GetServerSideProps` is of `IncomingMessage`, but when I try to pass it in as part of `NextPageContext` to `getSession`, it gets the session fine, but TS gives me ```Type 'IncomingMessage' is missing the following properties from type 'NextApiRequest': query, cookies, body, envts(2739)
```

**Comment by styxlab** [comment]
Hey @lluia! Thanks so much for this typescript package!

I just noticed that the `ContextProviderOptions` interface in `client.d.ts`
```
interface ContextProviderOptions {
    site?: string;
    basePath?: string;
    clientMaxAge?: number;
}
```
is missing the `keepalive` flag. Would it be possible to add it to the interface?

**Comment by JuanM04** [comment]
Hey @styxlab, see [this draft](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47310)

**Pull request merged**
